x/superfluid/keeper/grpc_query.go

```diff
@@ -278,6 +279,78 @@ func (q Querier) SuperfluidDelegationsByDelegator(goCtx context.Context, req *ty
 	return &res, nil
 }
 
+// UserSuperfluidPositionsPerConcentratedPoolBreakdown returns all the cl superfluid positions for the specified delegator in the specified concentrated liquidity pool.
+func (q Querier) UserSuperfluidPositionsPerConcentratedPoolBreakdown(goCtx context.Context, req *types.UserSuperfluidPositionsPerConcentratedPoolBreakdownRequest) (*types.UserSuperfluidPositionsPerConcentratedPoolBreakdownResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+
+	delAddr, err := sdk.AccAddressFromBech32(req.DelegatorAddress)
+	if err != nil {
+		return nil, err
+	}
+
+	// Get the position IDs for the specified pool ID and user address.
+	positions, err := q.Keeper.clk.GetUserPositions(ctx, delAddr, req.ConcentratedPoolId)
+	if err != nil {
+		return nil, err
+	}
+
+	// Query each position ID and determine if it has a lock ID associated with it, which implies the position is superfluid staked.
+	// Construct a response with the position ID, lock ID, the amount of cl shares staked, and what those shares are worth in staked osmo tokens.
+	var clPoolUserPositionRecords []types.ConcentratedPoolUserPositionRecord
+	for _, pos := range positions {
+		lockId, err := q.Keeper.clk.GetLockIdFromPositionId(ctx, pos.PositionId)
+		switch err.(type) {
+		case cltypes.PositionIdToLockNotFoundError:
+			continue
+		case nil:
+			if lockId == 0 {
+				continue
+			}
+
+			lock, err := q.Keeper.lk.GetLockByID(ctx, lockId)
+			if err != nil {
+				return nil, err
+			}
+
+			syntheticLock, err := q.Keeper.lk.GetSyntheticLockupByUnderlyingLockId(ctx, lockId)
+			if err != nil {
+				return nil, err
+			}
+
+			if syntheticLock.UnderlyingLockId == 0 {
+				continue
+			}
```

Comment:
q: this can happen when a regular position is just migrated. In that case, it has a regular but not a synth lock

Reply:
This case should technically not be possible, because we only link a lockId to a positionId IFF the lock is superfluid, and therefore any lockId that gets passed into `GetSyntheticLockupByUnderlyingLockId` should return a valid synthetic lock.

Reply:
Should we then error in this case instead of continuing?

Reply:
Yeah that is fine, added error here https://github.com/osmosis-labs/osmosis/pull/5600/commits/b8682d3db1c63d98f04e855f6d195d1bcdcea2b2